### PR TITLE
C19288 accord async reads are unsafe

### PR DESCRIPTION
**Pull request by bdeggleston**

### CODE REVIEW

File: accord-core/src/main/java/accord/messages/ReadData.java [hunk 1]
```diff
@@ -359,11 +359,13 @@ void cancel(@Nullable SafeCommandStore safeStore)
             if (safeCommand != null) safeCommand.removeListener(this);
             waitingOn.clear(safeStore.commandStore().id());
         }
+
         // TODO (expected): efficient unsubscribe mechanism
-        node.commandStores().mapReduceConsume(this, waitingOn.stream(), forEach(in -> {
-            SafeCommand safeCommand = in.ifInitialised(txnId);
-            if (safeCommand != null) safeCommand.removeListener(this);
-        }, node.agent()));
+        if (waitingOn != null)
```
Comment (dcapwell) [inline accord-core/src/main/java/accord/messages/ReadData.java:364]: did you hit a NPE?  since we touch `waitingOn.stream()` we would hit a NPE.

**Review by dcapwell — COMMENTED** [review]
am I right that async reads are fine, its just async ephemeral reads is where we have issues?  Else I don't see the real change that fixes async reads

File: accord-core/src/main/java/accord/messages/ReadData.java (hunk 1 above)
Comment (bdeggleston) [reply accord-core/src/main/java/accord/messages/ReadData.java:364]: yes, if there are multiple calls to cancel it will throw an NPE

**Comment by bdeggleston** [comment]
I think that async reads are safe by virtue of the execution process, but the assertion was assuming they were synchronous. Was there another issue with ephemeral reads? I'm less familiar with them, but it seems like it would be ok for an ephemeral read to read the from "future", so long as they don't go backwards in time

**Comment by dcapwell** [comment]
We disable Ephemeral Reads in C* as they don't work (let alone a single cache evict will cause chaos), so thats fine by me.

**Pull request closed without merging**